rl4co/envs/scheduling/jssp.py

```diff
@@ -0,0 +1,665 @@
+
+    Args:
+
+    Note:
+        -
+    """
+
+    name = "jssp"
+
+    def __init__(self, num_jobs, num_machines, **kwargs):
+        super().__init__(**kwargs)
+        self.batch_size = torch.Size([1])
+        adjacency_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size((1, num_jobs * num_machines, num_jobs * num_machines)),
+            device=self.device,
+            dtype=torch.int64,
+        )
+        features_spec = UnboundedContinuousTensorSpec(
+            shape=torch.Size((1, num_jobs * num_machines, 2)),
+            device=self.device,
+        )
+        feasible_actions_spec = DiscreteTensorSpec(
+            n=num_jobs * num_machines,
+            shape=torch.Size(
+                (
+                    1,
+                    num_jobs,
+                )
+            ),
+            device=self.device,
+            dtype=torch.int64,
+        )
+        action_mask_spec = DiscreteTensorSpec(
+            n=2,
+            shape=torch.Size(
+                (
+                    1,
+                    num_jobs,
+                )
+            ),
+            device=self.device,
+            dtype=torch.bool,
+        )
+
+        self.observation_spec = CompositeSpec(
+            adjacency=adjacency_spec,
+            features=features_spec,
+            feasible_actions=feasible_actions_spec,
+            action_mask=action_mask_spec,
+            shape=self.batch_size,
+        )
+
+        self.action_spec = DiscreteTensorSpec(
+            n=num_jobs,
+            shape=self.batch_size,
+            device=self.device,
+            dtype=torch.int64,
+        )
+
+        self.num_jobs = num_jobs
+        self.num_machines = num_machines
+        self.num_tasks = self.num_jobs * self.num_machines
+        # the task id for first column
+        self.first_col = torch.arange(
+            start=0,
+            end=self.num_tasks,
+            step=num_machines,
+            dtype=torch.int64,
+            device=self.device,
+        )
+        # the task id for last column
+        self.last_col = torch.arange(
+            start=num_machines - 1,
+            end=self.num_tasks,
+            step=num_machines,
+            dtype=torch.int64,
+            device=self.device,
+        )
+
+        # initialize zero matrices for memory allocation
+        self.machines = torch.zeros(
+            (self.num_jobs, self.num_machines),
+            dtype=torch.float32,
+            device=self.device,
+        )
+        self.durations = torch.zeros(
+            (self.num_jobs, self.num_machines),
+            dtype=torch.float32,
+            device=self.device,
+        )
+        self.durations_cp = torch.zeros(
+            (self.num_jobs, self.num_machines),
+            dtype=torch.float32,
+            device=self.device,
+        )
+
+    def done(self):
+        if len(self.partial_sol_sequence) == self.num_tasks:
+            return torch.tensor(True)
+        return torch.tensor(False)
```

Comment:
Important note:
In general, whenever there is a `torch.tensor(...)`, `torch.ones(...)` etc to be instantiated , it is better to do so on the device. Since we use GPU for training and we can move the envs there for offloading the CPU, it would be best to instantiate directly there.
For instance:

```python
...
return torch.tensor(False, device=self.device)
```

The same applies to other tensors instantiated this way - at the moment, trying to set the env on GPU raises some device mismatch error, but it can be easily fixed :)